src/Workspaces/Core/Portable/Workspace/Solution/ProjectDependencyGraph.cs

```diff
@@ -14,35 +14,288 @@ namespace Microsoft.CodeAnalysis
+            ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> existingTransitiveReferencesMap,
+            ProjectId projectId,
+            IReadOnlyList<ProjectId> referencedProjectIds)
+        {
+            // To update our forward transitive map, we need to add referencedProjectIds (and their transitive dependencies) to the transitive references
+            // of projects. First, let's just compute the new set of transitive references. It's possible while doing so we'll discover that we don't
+            // know the transitive project references for one of our new references. In that case, we'll use null as a sentinel to mean "we don't know" and
+            // we propogate the not-knowingness. But let's not worry about that yet. First, let's just get the new transitive reference set.
+            var newTransitiveReferences = new HashSet<ProjectId>(referencedProjectIds);
+
+            foreach (var referencedProjectId in referencedProjectIds)
+            {
+                if (existingTransitiveReferencesMap.TryGetValue(referencedProjectId, out var additionalTransitiveReferences))
+                {
+                    newTransitiveReferences.UnionWith(additionalTransitiveReferences);
+                }
+                else
+                {
+                    newTransitiveReferences = null;
+                    break;
+                }
+            }
+
+            // We'll now loop through each entry in our existing cache and compute updates. We'll accumulate them into this builder.
+            var builder = existingTransitiveReferencesMap.ToBuilder();
+
+            foreach (var projectIdToUpdate in existingTransitiveReferencesMap.Keys)
+            {
+                existingTransitiveReferencesMap.TryGetValue(projectIdToUpdate, out var existingTransitiveReferences);
+
+                // The projects who need to have their caches updated are projectIdToUpdate (since we're obviously updating it!)
+                // and also anything that depended on us.
+                if (projectIdToUpdate == projectId || existingTransitiveReferences?.Contains(projectId) == true)
+                {
+                    // This needs an update. If we know what to include in, we'll union it with the existing ones. Otherwise, we don't know
+                    // and we'll remove any data from the cache.
+                    if (newTransitiveReferences != null && existingTransitiveReferences != null)
+                    {
+                        builder[projectIdToUpdate] = existingTransitiveReferences.Union(newTransitiveReferences);
+                    }
+                    else
+                    {
+                        // Either we don't know the full set of the new references being added, or don't know the existing set projectIdToUpdate.
+                        // In this case, just remove it
+                        builder.Remove(projectIdToUpdate);
+                    }
+                }
+            }
```

Comment:
this loop is subtle.  but i think it's correct.  basically because you'll go back and retouch anyone who currently thinks they depend on.   if they think tey depend on you, you either update them with all the new transitive references.  Or, if you had to bail out, you actually backpropagate the idea of them 'not knowing' what hte transitive referneces are.

I feel ok about this algorithm.  but tests of lots of interesting scenarios will give me confidence.